src/legacy/shrink.rs

```diff
@@ -0,0 +1,426 @@
+use std::collections::VecDeque;
+use std::io::{self, copy, Read};
+
+use thiserror::Error;
+
+use super::bitstream::{lsb, BitStream};
+
+const MIN_CODE_SIZE: u8 = 9;
+const MAX_CODE_SIZE: u8 = 13;
+
+const MAX_CODE: usize = (1 << MAX_CODE_SIZE) - 1;
+const INVALID_CODE: u16 = u16::MAX;
+const CONTROL_CODE: usize = 256;
+const INC_CODE_SIZE: u64 = 1;
+const PARTIAL_CLEAR: u64 = 2;
+
+// const HASH_BITS: usize = MAX_CODE_SIZE + 1; /* For a load factor of 0.5. */
+// const HASHTAB_SIZE: usize = 1 << HASH_BITS;
+const UNKNOWN_LEN: u16 = u16::MAX;
+
+#[derive(Error, Debug)]
+enum ShrinkError {
+    #[error("self-referential code")]
+    InvalidPrefixCode,
+
+    #[error("first code needs to be literal")]
+    FirstCodeNeedsToBeLiteral,
+
+    #[error("invalid code")]
+    InvalidCode,
+
+    #[error("prev code no longer valid")]
+    PrevCodeNoLongerValid,
+}
+struct CodeQueue {
+    next_idx: usize,
+    codes: [u16; MAX_CODE as usize - CONTROL_CODE + 1],
+}
+
+impl CodeQueue {
+    fn new() -> Self {
+        let mut codes = [0; MAX_CODE as usize - CONTROL_CODE + 1];
+        for (i, code) in (CONTROL_CODE as u16 + 1..=MAX_CODE as u16).enumerate() {
+            codes[i] = code;
+        }
+        Self { next_idx: 0, codes }
+    }
+
+    // Return the next code in the queue, or INVALID_CODE if the queue is empty.
+    fn next(&self) -> u16 {
+        //   assert(q->next_idx < sizeof(q->codes) / sizeof(q->codes[0]));
+        self.codes[self.next_idx]
+    }
+
+    /// Return and remove the next code from the queue, or return INVALID_CODE if
+    /// the queue is empty.
+    fn remove_next(&mut self) -> u16 {
+        let code = self.next();
+        if code != INVALID_CODE {
+            self.next_idx += 1;
+        }
+        code
+    }
+}
+
+#[derive(Clone, Copy)]
+struct Codetab {
+    prefix_code: u16, // INVALID_CODE means the entry is invalid.
+    ext_byte: u8,
+    len: u16,
+    last_dst_pos: usize,
+}
```

Comment:
```suggestion
#[repr(packed)]
struct Codetab {
    last_dst_pos: usize,
    prefix_code: Option<u16>,
    ext_byte: u8,
    len: u16
}
```
This should fit into 128 bits on 64-bit platforms and 96 bits on 32-bit platforms, all with no penalty for an unaligned access.

Reply:
Rust doesn't like that - I get errors. Why repr(packed) ?

Reply:
It prevents any alignment padding. Usually scalar values are aligned on their size, and arrays on their element size.